pkg/controllers/externalsecret/externalsecret_controller.go

```diff
@@ -245,228 +307,353 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 		case esv1beta1.DeletionPolicyDelete:
 			// safeguard that we only can delete secrets we own
 			// this is also implemented in the es validation webhook
-			if externalSecret.Spec.Target.CreationPolicy != esv1beta1.CreatePolicyOwner {
-				err := fmt.Errorf(errInvalidCreatePolicy, externalSecret.Spec.Target.CreationPolicy)
-				r.markAsFailed(log, errDeleteSecret, err, &externalSecret, syncCallsError.With(resourceLabels))
+			creationPolicy := externalSecret.Spec.Target.CreationPolicy
+			if creationPolicy != esv1beta1.CreatePolicyOwner {
+				err := fmt.Errorf(errDeleteCreatePolicy, secretName, creationPolicy)
+				r.markAsFailed(msgErrorDeleteSecret, err, externalSecret, syncCallsError.With(resourceLabels))
 				return ctrl.Result{}, err
 			}
 
-			if err := r.Delete(ctx, secret); err != nil && !apierrors.IsNotFound(err) {
-				r.markAsFailed(log, errDeleteSecret, err, &externalSecret, syncCallsError.With(resourceLabels))
-				return ctrl.Result{}, err
+			// delete the secret, if it exists
+			if existingSecret.UID != "" {
+				if err := r.Delete(ctx, existingSecret); err != nil && !apierrors.IsNotFound(err) {
+					r.markAsFailed(msgErrorDeleteSecret, err, externalSecret, syncCallsError.With(resourceLabels))
+					return ctrl.Result{}, err
+				}
 			}
 
-			conditionSynced := NewExternalSecretCondition(esv1beta1.ExternalSecretReady, v1.ConditionTrue, esv1beta1.ConditionReasonSecretDeleted, "secret deleted due to DeletionPolicy")
-			SetExternalSecretCondition(&externalSecret, *conditionSynced)
-			return ctrl.Result{RequeueAfter: refreshInt}, nil
+			r.markAsDone(externalSecret, start, log, esv1beta1.ConditionReasonSecretDeleted, msgDeleted)
+			return r.getRequeueResult(externalSecret), nil
 		// In case provider secrets don't exist the kubernetes secret will be kept as-is.
 		case esv1beta1.DeletionPolicyRetain:
-			r.markAsDone(&externalSecret, start, log)
-			return ctrl.Result{RequeueAfter: refreshInt}, nil
+			r.markAsDone(externalSecret, start, log, esv1beta1.ConditionReasonSecretSynced, msgSyncedRetain)
+			return r.getRequeueResult(externalSecret), nil
 		// noop, handled below
 		case esv1beta1.DeletionPolicyMerge:
 		}
 	}
 
-	mutationFunc := func() error {
+	// mutationFunc is a function which can be applied to a secret to make it match the desired state.
+	mutationFunc := func(secret *v1.Secret) error {
```

Comment:
Can you put this into a function please and out of the reconcile loop? It's not significantly larger so I think it should be fine to put it outside.

Reply:
we went through this yesterday - I don't think this is possible because the mutation function depends on several internal variables of the main loop. I think this improvement could possibly be tackled out of this PR maybe? 🤔

Reply:
Okay 👍

Reply:
Let's create a follow up.